### PR TITLE
Add option nounset to default to strict parsing.

### PR DESCRIPTION
**Pull request by sayanarijit**
`nounset=True` will behave like bash's `set -u` or `set -o nounset`
i.e. when any variable is undefined, it will raise `UnboundVariable`
exception.

This is useful where using the `${VAR:?}` syntax is not an option.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -39,6 +41,42 @@ Examples
...
+```
+
+> NOTE: Another way is to use the `${VAR?}` or `${VAR:?}` syntax. See the examples in tests.
+
+### export RECOVER_NULL="foo"
+
+If you want to temporarily disable strict parsing both for `nounset=True` and the `${VAR:?}` syntax, set environment variable `RECOVER_NULL=somevalue`.
+
+```bash
+export RECOVER_NULL=foo
```
Comment (mnieber) [inline README.md:65]: Doing this with an environment variable seems very dangerous. If you have a pipeline that involves your own tool but also someone else's tool that happens to use expandvars, then this other tool can start behaving wrong.

Comment (sayanarijit) [reply README.md]: Hi. Valid point too. We shouldn't use `export`. But this feature comes in handy so often that I'd still keep it.
I have added WARNINGs to remind the consequences of using `export`.

Comment (mnieber) [reply README.md]: The problem is that some of the people who would like to see the warning won't be able to see it. If Alice and Bob each create a tool, and I put them both in my stack, then Bob's tool can break when Alice's tool has a script that says "RECOVER_ALL=foo". I'm just an unsuspecting user who would be bitten by this (I probably don't even know that the tools I am relying on use expandvars internally).

Comment (sayanarijit) [reply README.md]: Sorry I misspelled it in the first PR (corrected it now). The name is actually `EXPANDVARS_RECOVER_NULL`. I doubt someone will use this name unintentionally.

**Comment by mnieber** [comment]
I think you misunderstood my point. If you have a toolchain where 3
different tools are using expandvars internally, then the toolchain
will only work correctly if these 3 tools have the same expectation
about EXPANDVARS_RECOVER_NULL.
For example, if these 3 tools don't do anything
with EXPANDVARS_RECOVER_NULL but you add a fourth tool that includes a
script (sourced when the toolchain is brought up) which
intentionally sets EXPANDVARS_RECOVER_NULL then it will break the other 3
tools.

**Comment by sayanarijit** [comment]
In that case, I think it will be a fault on the tool developer side. Just like it's a common knowledge that we shouldn't update `PATH` and other critical environment variables in a script without knowing what we're doing, it's the same here. This variable must be used while running the whole toolchain. Not individual tools. For individual tools, if bypassing strictness is needed and using this variable seems to be the only way, that tool should set the variable in a temporary manner. Without exporting it. Just like we'd temporarily update PATH.

**Comment by sayanarijit** [comment]
If you think I misunderstood again, could you please give a real world example? Or a real world scenario? It'll be easier that way.

**Comment by mnieber** [comment]
Sure! If I'm developing the tool chain and I am completely unaware of
expandsvars (don't even know that such a package exists) and I use a bunch
of 3rd party tools,
then I won't know which ones are using expandvars. I might add a new tool
that sets EXPANDVARS_RECOVER_NULL and  breaks some of the other tools that
also use expandvars.
All I see as a developer of the toolchain is that things mysteriously
started to break just by adding a tool.

**Comment by sayanarijit** [comment]
> I might add a new tool that sets EXPANDVARS_RECOVER_NULL 

If that toos sets the variable locally, it shouldn't affect the other tools. If it exports the variable globally, that tool might also be exporting critical variables like PATH and USER globally. You probably should avoid using anything that depends on this tool.

**Comment by mnieber** [comment]
| You probably should avoid using anything that depends on this tool.

How can I find out what depends on this tool?

**Comment by sayanarijit** [comment]
If adding a tool breaks my system, it means it's "incompatible" with my system. Then I either find a compatible alternative, or investigate why is it incompatible. If after using a tool I see it's messing with my system global variables, I'd be uncomfortable using that tool. On the other hand, it's that tool has to set global environment variable like tmux sets "TMUX" and "TMUX_PANE" and python depends on "PYTHON_PATH" and "PYTHON_HOME", there's a best practice to prefix the variable name properly like "EXPANDVARS_*". If it still sets "EXPANDVARS_RECOVER_NULL" for some other purpose, then it's just an   incompatibility issue I don't see ever happening.

And if is sets "EXPANDVARS_RECOVER_NULL" intentionally to influence other tools, Just like [direnv](https://github.com/direnv/direnv) and [nix](https://github.com/NixOS/nix) does, it should know what it's doing. It's the responsibility of that tool's developer to make it compatible with other tools.

**Comment by mnieber** [comment]
You said it yourself (not me!): You probably should avoid using anything that depends on this tool.
Then I asked "How can I find out what depends on this tool?" and (as you realize) this question is impossible to answer.
Sorry, I will leave it here. I do like the package though, very useful.

**Comment by sayanarijit** [comment]
Thanks 😄 ...

Sorry I can't give a better answer than that. This little feature adds a lot of convenience at the cost an incompatibility issue that's very unlikely to happen. Hence, I'm fine with the trade-off.

**Comment by mnieber** [comment]
Okay, if that's a risk you are willing to take (I think if people get
bitten by this, it could be quite bad)

**Comment by sayanarijit** [comment]
Alright. I'll create a separate issue to see what other users of this library feel about this.

**Pull request merged**